### PR TITLE
Improve phi assignment by picking an earlier/smaller set of nodes for SetPhiStmt

### PR DESCRIPTION
**Pull request by zbanks**
In `assign_phis()`, rather than emit the `SetPhiStmt` at the end of each of parents, try to find a different set of nodes that has equivalent behavior, but will likely produce nicer code.

- Prefer picking *earlier* nodes when possible. This helps avoid getting assignments inside complex conditionals (see `andor` tests), and can result in similar phi nodes "bunching up" in a way that's easier to refactor (see `phi_a0_2`-`5` in `multi-switch` test).
- If two or more of `phi.node.parents` share a dominator, we can do the assignment *once* in the dominator, as long as the value of the register at that node is the same. This reduces the number of statements.
- I originally thought loops would be tricky. I was worried we needed to avoid moving phi assignments out of loops, because that's not equivalent. I found some places where the phis got moved out of loops, but it turned out that moving the phi out made it *more* correct. ([`Mio0_Decompress`, in `mio0.c`](https://gist.github.com/zbanks/d628c6a4bfb3c9bfec8d4246ba6ac4bb)) I couldn't manage to reproduce this behavior in an end-to-end test.

### CODE REVIEW

File: src/translate.py [hunk 1]
```diff
@@ -3168,11 +3197,17 @@ def assign_phis(used_phis: List[PhiExpr], stack_info: StackInfo) -> None:
         phi = used_phis[i]
         assert phi.num_usages > 0
         assert len(phi.node.parents) >= 2
-        exprs = []
+
+        # Group parent nodes by the value of their phi register
+        equivalent_nodes: Dict[Expression, List[Node]] = {}
```
Comment (simonlindholm) [inline src/translate.py:3202]: `= defaultdict(list)` if you want to avoid the `if expr not in equivalent_nodes` check

File: tests/end_to_end/andor_assignment/irix-o2-out.c [hunk 2]
```diff
@@ -62,9 +62,7 @@ s32 test(s32 arg0, s32 arg1, s32 arg2, s32 arg3) {
             phi_v1_3 = phi_v1_6 + 5;
         }
     }
-    phi_v1_4 = phi_v1_3;
-    phi_v1_7 = phi_v1_3;
-    if ((phi_s0_2 != 0) && (phi_t0_2 != 0) && (sp24 = phi_v1_3, (func_00400090(phi_s0_2 + phi_t0_2) != 0)) && (arg3 != 0)) {
+    if ((phi_s0_2 != 0) && (phi_t0_2 != 0) && (sp24 = phi_v1_3, phi_v1_4 = phi_v1_3, phi_v1_7 = phi_v1_3, (func_00400090(phi_s0_2 + phi_t0_2) != 0)) && (arg3 != 0)) {
```
Comment (simonlindholm) [inline tests/end_to_end/andor_assignment/irix-o2-out.c:65]: do you know why this change happened? I don't understand it

**Review by simonlindholm — APPROVED** [review]

File: tests/end_to_end/andor_assignment/irix-o2-out.c (hunk 2 above)
Comment (zbanks) [reply tests/end_to_end/andor_assignment/irix-o2-out.c:65]: The function call `func_00400090` saves/restores the register `v1`. [When the register is restored, it is wrapped in an `EvalOnceExpr`](https://github.com/matt-kempster/mips_to_c/blob/master/src/translate.py#L3421-L3435) which is  no longer unwrapped.
